### PR TITLE
#37 fix throwing exception in case of unrecognized first token

### PR DESCRIPTION
**Pull request by Phill101**

### CODE REVIEW

File: modules/jackson-backend/src/main/scala/tethys/jackson/JacksonTokenIterator.scala [hunk 1]
```diff
@@ -1,12 +1,19 @@
...
 import tethys.readers.tokens.{BaseTokenIterator, TokenIterator}
 
 class JacksonTokenIterator(jsonParser: JsonParser) extends BaseTokenIterator {
-  private[this] var token: Token = fromId(jsonParser.currentTokenId())
+  private[this] var token: Token = {
+    try {
+      jsonParser.nextToken() // move parser to first token
+      fromId(jsonParser.currentTokenId())
+    } catch {
+      case _: JsonParseException => UnrecognizedStart
```
Comment (eld0727) [inline modules/jackson-backend/src/main/scala/tethys/jackson/JacksonTokenIterator.scala]: I think that is a wrong behavior.
JsonParseException could be thrown in any position, but it's handled only here
And new type of token :(
Maybe it's better to fix this issue in `tethys.ReaderReaderOps` functions?

Comment (eld0727) [reply modules/jackson-backend/src/main/scala/tethys/jackson/JacksonTokenIterator.scala]: Another option is to change signature of `tethys.readers.tokens.TokenIteratorProducer#fromReader` from `Reader => TokenIterator` to `Reader => Either[ReaderError, TokenIterator]`

File: modules/core/src/main/scala/tethys/package.scala [hunk 2]
```diff
@@ -44,7 +44,8 @@ package object tethys {
 
   implicit class ReaderReaderOps(val reader: Reader) extends AnyVal {
     def readJson[A](implicit jsonReader: JsonReader[A], producer: TokenIteratorProducer): Either[ReaderError, A] = {
-      producer.fromReader(reader).readJson[A]
+      implicit val root: FieldName = FieldName()
+      ReaderError.catchNonFatal(producer.fromReader(reader)).right.flatMap(_.readJson[A])
```
Comment (Phill101) [inline modules/core/src/main/scala/tethys/package.scala]: This look much better, but still slightly weird.
I Explore your suggestion to change fromReader signature but it's seems to be a little bit complicated :D

File: modules/core/src/main/scala/tethys/commons/TokenNode.scala [hunk 3]
```diff
@@ -100,36 +100,37 @@ object TokenNode {
...
-          else if (token.isFieldName) FieldNameNode(iterator.fieldName())
-          else if (token.isStringValue) StringValueNode(iterator.string())
-          else if (token.isNumberValue) iterator.number() match {
-            case v: java.lang.Short => ShortValueNode(v)
-            case v: java.lang.Integer => IntValueNode(v)
-            case v: java.lang.Long => LongValueNode(v)
-            case v: java.lang.Float => FloatValueNode(v)
-            case v: java.lang.Double => DoubleValueNode(v)
-            case n => NumberValueNode(n)
+      json.toTokenIterator.right.map { iterator =>
```
Comment (eld0727) [inline modules/core/src/main/scala/tethys/commons/TokenNode.scala]: This method is used only for test reason, so i think we can just throw an exception

Comment (Phill101) [reply modules/core/src/main/scala/tethys/commons/TokenNode.scala]: done

**Pull request merged**